### PR TITLE
Add Polytype/Monotype Distinction

### PR DESCRIPTION
**Pull request by joshrule**

### CODE REVIEW

**Comment by lucasem** [comment]
Key points from our discussion offline, for the sake of public record-keeping:
- `Polytype` → `TypeSchema`
- Make consistent the terminology of substitution/binding.
- Distinguish `Variable` and `MetaVariable`/`QuantifiedVariable`.
  - We may want `TypeSchema` to be an isolated enum that doesn't rely on `Type` so we can make this distinction work nicely. This would probably make us want a more complex `ptp!` macro that behaves like `tp!`, and a corresponding `parrow!` macro.
- `ptp!` macro with fancier ways of supplying many quantified variables, e.g. `ptp!{(0, 1) => ptp!(parrow![ptp(0), ptp(1)])}` or something analogous.
- `as_arrow` method for a `Type` which gives us `Option<(&Type, &Type)>` (the arrow's arg and ret).
- More methods on `TypeSchema`, like getting a list of bound quantified variables.

**Comment by joshrule** [comment]
> - Distinguish Variable and MetaVariable/QuantifiedVariable.
>   - We may want TypeSchema to be an isolated enum that doesn't rely on Type so we can make this distinction work nicely. This would probably make us want a more complex ptp! macro that behaves like tp!, and a corresponding parrow! macro.

Thinking about this a bit more this morning, I don't know that the `TypeVariable`/`MetaTypeVariable` distinction is the right one. Quantified type variables are still plain-old type variables. Instantiation has been working fine since I fixed the conceptual bug of adding quantified variables to a context. The main issue now issue, I think, is providing tools to avoid variable capture when creating TypeSchemes. Perhaps the `ptp!` macro should allow you to specify somewhere which variables in a monotype should be considered free and which bound?

File: src/lib.rs [hunk 1]
```diff
@@ -175,41 +315,72 @@ pub enum Type {
...
+            Type::Constructed("→", ref args) => {
+                let mut tps = args[1].args().unwrap_or_default();
+                tps.push_front(&args[0]);
+                Some(tps)
+            }
+            Type::Variable(_) | Type::Constructed(..) => None,
+        }
+    }
+    /// Optionally return the return type of an arrow.
+    pub fn returns(&self) -> Option<&Type> {
```
Comment (lucasem) [inline src/lib.rs]: Maybe this and the preceding method (`args()`) should be replaced with something like:

```rust
/// Like [`as_arrow`], but recursively gets all curried argument types.
pub fn deconstruct_arrow(&self) -> Option<(VecDeque<&Type>, &Type)> { .. }
```

File: src/lib.rs [hunk 2]
```diff
@@ -226,115 +397,121 @@ impl Type {
...
-                Type::Arrow(Box::new(Arrow { arg, ret }))
-            }
             Type::Constructed(name, ref args) => {
                 let args = args.iter().map(|t| t.apply(ctx)).collect();
                 Type::Constructed(name, args)
             }
             Type::Variable(v) => {
-                if let Some(tp) = ctx.substitutions.get(&v) {
+                if let Some(tp) = ctx.substitution.get(&v) {
+                    // hmm... is this right?
```
Comment (lucasem) [inline src/lib.rs]: 🤔

File: src/lib.rs [hunk 3]
```diff
@@ -582,24 +698,20 @@ impl Context {
...
+    /// which it's called should be discarded if there's an error.
     fn unify_internal(&mut self, t1: &Type, t2: &Type) -> Result<(), UnificationError> {
         let t1 = t1.apply(self);
         let t2 = t2.apply(self);
         if t1 == t2 {
             return Ok(());
         }
-        if !t1.is_polymorphic() && !t2.is_polymorphic() {
-            return Err(UnificationError::Failure(t1, t2));
-        }
```
Comment (lucasem) [inline src/lib.rs]: We should keep this early return (using `t.free_vars().is_empty()`)

File: tests/tests.rs [hunk 4]
```diff
@@ -264,44 +227,28 @@ fn test_unify_both_sides_polymorphic_occurs() {
...
+//     assert_eq!(&bindings[&3], &tp!(1));
+//     // like replaces like
+//     assert_eq!(variables(&t1), variables(&t2));
+//     // substitutions are not made
+//     assert_eq!(t3, tp!(list(tp!(1))));
+//     // context is updated
+//     assert_eq!(ctx.new_variable(), tp!(2));
+//     assert_eq!(ctx.substitutions().get(&1).unwrap(), &dummy);
+//     assert_eq!(ctx.substitutions().len(), 1);
+// }
```
Comment (lucasem) [inline tests/tests.rs]: This should be uncommented and revised if necessary

**Comment by lucasem** [comment]
Actually, now that I've thought more about it, forget about the early return in `unify_internal`. Unification should be cheap enough that it doesn't actually help.

File: src/lib.rs [hunk 5]
```diff
@@ -644,45 +751,70 @@ mod parser {
...
+                                     constructed_simple)) >>
+                     alt!(tag!("→") | tag!("->")) >>
+                     beta: ws!(monotype) >>
+                     (Type::arrow(alpha, beta)))
+    );
+    named!(parenthetical<CompleteStr, Type>,
+           delimited!(tag!("("), arrow, tag!(")"))
+    );
+    named!(binding<CompleteStr, TypeSchema>,
+           do_parse!(opt!(tag!("∀")) >>
```
[the remark this answers is not in the document]
Comment (joshrule) [reply src/lib.rs]: It's already ASCII friendly in that the `∀` is optional. You can write `∀t0. t0 → t0` or `t0. t0 -> t0`.

File: src/lib.rs (hunk 1 above)
Comment (joshrule) [reply src/lib.rs]: Maybe, though my intuition is that I often have use for one or the other. Let's leave it for now and change later as use dictates.

**Comment by lucasem** [comment]
Great work! Thanks again for doing this!

**Pull request merged**